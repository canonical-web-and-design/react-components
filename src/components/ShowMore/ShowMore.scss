@import "~vanilla-framework/scss/settings";

.p-show-more__container {
  align-items: center;
  display: grid;
  grid-template-columns: 1fr auto 0.5rem;
}

.p-show-more__container::before,
.p-show-more__container::after {
  border-bottom: 1px dotted $color-mid-light;
  content: "";
  height: 0;
}

.p-show-more__button {
  margin: 0 0.5rem;
  padding: 0 0.5rem;
}
