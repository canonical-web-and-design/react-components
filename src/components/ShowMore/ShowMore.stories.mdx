import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";
import { useState } from "react";

import ShowMore from "./ShowMore";
import ModularTable from "../ModularTable";

<Meta title="ShowMore" component={ShowMore} />

## ShowMore

By default component renders as a button with a label.

### Props

<ArgsTable of={ShowMore} />

### Default

export const doNothing = () => {};

<Canvas>
  <Story name="Default">
    <ShowMore label="Show more items" onClick={doNothing} />
  </Story>
</Canvas>

### Loading

When data is loaded asynchronusly `isLoading` prop should be set to `true` to shows a spinner animation.

<Canvas>
  <Story name="Loading">
    <ShowMore label="Show more items" isLoading onClick={doNothing} />
  </Story>
</Canvas>


### Complete example with ModularTable

Example below shows a basic `ModularTable` with `ShowMore` component below.

export const getRandomData = () => [
  { data: Math.random() * 1000 },
  { data: Math.random() * 1000 },
  { data: Math.random() * 1000 },
  { data: Math.random() * 1000 },
  { data: Math.random() * 1000 },
]

<Canvas>
  <Story name="Click event">
    {() => {
      const [data, setData] = useState(getRandomData());
      const hasMore = data.length < 20;
      const loadMore = () => {
        setData(data.concat(getRandomData()));
      };
      const columns = [
        {
          Header: "Id",
          accessor: (d, i) => i,
        },
        {
          Header: "Data",
          accessor: "data",
        }
      ];
      return (
        <>
          <ModularTable data={data} columns={columns} />
          { hasMore &&
            <ShowMore label="Load more" onClick={loadMore} />
          }
        </>
      )
    }}
  </Story>
</Canvas>
