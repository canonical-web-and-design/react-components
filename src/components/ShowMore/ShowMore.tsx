import React, { SyntheticEvent } from "react";
import PropTypes from "prop-types";

import "./ShowMore.scss";

type Props = {
  label: string;
  onClick: (event: SyntheticEvent) => void;
  summary?: string;
  isLoading?: boolean;
};

const ShowMore = ({
  isLoading,
  summary,
  label,
  onClick,
}: Props): JSX.Element => {
  // prevent browser from changing the URL
  const handleClick = (event: SyntheticEvent) => {
    event.preventDefault();
    onClick(event);
  };

  return (
    <span>
      {summary && (
        <>
          <span>{summary}</span>{" "}
        </>
      )}
      {isLoading ? (
        <i className="p-icon--spinner u-animation--spin" />
      ) : (
        <a href="#show-more" onClick={handleClick}>
          {label}
        </a>
      )}
    </span>
  );
};

ShowMore.propTypes = {
  summary: PropTypes.string,
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
};

export default ShowMore;
