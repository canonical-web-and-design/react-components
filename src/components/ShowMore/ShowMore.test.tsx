import { shallow } from "enzyme";
import React from "react";

import ShowMore from "./ShowMore";

describe("<ShowMore />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    const component = shallow(
      <ShowMore label="Show more" onClick={() => false} />
    );

    expect(component).toMatchSnapshot();
  });

  // unit tests
  it("renders a spinner icon", () => {
    const component = shallow(
      <ShowMore label="Show more" onClick={() => false} isLoading />
    );

    expect(
      component.find("i").first().hasClass("p-icon--spinner")
    ).toBeTruthy();
  });

  it("can handle button click events", () => {
    const onClick = jest.fn();
    const component = shallow(<ShowMore label="Show more" onClick={onClick} />);

    component.find("button").first().simulate("click");
    expect(onClick).toHaveBeenCalled();
  });
});
