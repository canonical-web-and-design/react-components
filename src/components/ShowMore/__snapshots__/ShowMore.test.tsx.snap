// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShowMore /> renders and matches the snapshot 1`] = `
<small
  className="u-text--muted"
>
  <span>
    Showing some items
  </span>
   
  <a
    href="#show-more"
    onClick={[Function]}
  >
    Show more
  </a>
</small>
`;
